pkg/controller/baremetalhost/baremetalhost_controller.go

```diff
@@ -133,6 +133,68 @@ func (info *reconcileInfo) publishEvent(reason, message string) {
 	info.events = append(info.events, info.host.NewEvent(reason, message))
 }
 
+// hostConfigData is an implementation of host configuration data interface.
+// Object is able to retrive data from secrets referenced in a host spec
+type hostConfigData struct {
+	host   *metal3v1alpha1.BareMetalHost
+	log    logr.Logger
+	client client.Client
+}
+
+// Generic method for data extraction from a Secret. Function uses dataKey
+// parameter to detirmine which data to return in case secret contins multiple
+// keys
+func (hcd *hostConfigData) getSecretData(name, namespace, dataKey string) ([]byte, error) {
+	secret := &corev1.Secret{}
+	key := types.NamespacedName{
+		Name:      name,
+		Namespace: namespace,
+	}
+	if err := hcd.client.Get(context.TODO(), key, secret); err != nil {
+		errMsg := fmt.Sprintf("failed to fetch user data from secret %s defined in namespace %s", name, namespace)
+		return nil, errors.Wrap(err, errMsg)
+	}
+
+	data, ok := secret.Data[dataKey]
+	if !ok {
+		hostConfigDataError.WithLabelValues(dataKey).Inc()
+		return nil, NoDataInSecretError{secret: name, key: dataKey}
+	}
+
+	return data, nil
+}
+
+// UserData get Operating System configuration data
+func (hcd *hostConfigData) UserData() (string, error) {
+	if hcd.host.Spec.UserData == nil {
+		hcd.log.Info("UserData is not set return empty string")
+		return "", nil
+	}
+	userData, err := hcd.getSecretData(
+		hcd.host.Spec.UserData.Name,
+		hcd.host.Spec.UserData.Namespace,
+		"userData",
+	)
+	if err != nil {
+		return "", err
+	}
+	return string(userData), nil
+
+}
+
+// NetworkData get network configuration
+func (hcd *hostConfigData) NetworkData() ([]byte, error) {
+	if hcd.host.Spec.NetworkData == nil {
+		hcd.log.Info("NetworkData is not set returning epmty(nil) data")
+		return nil, nil
```

Comment:
Why nil here but an empty string for user data?

Reply:
Network data now returns 'string' so this is updated in latest commit